core/services/ocr2/plugins/ocr2keeper/evm21/mercury/streams/streams.go

```diff
@@ -0,0 +1,330 @@
+		wg.Add(1)
+		func(i int, lookup *mercury.StreamsLookup) {
+			s.threadCtrl.Go(func(ctx context.Context) {
+				s.doLookup(ctx, &wg, lookup, i, checkResults)
+			})
+		}(i, lookup)
+	}
+	wg.Wait()
+
+	// don't surface error to plugin bc StreamsLookup process should be self-contained.
+	return checkResults
+}
+
+// buildResult checks if the upkeep is allowed by Mercury and builds a streams lookup request from the check result
+func (s *streams) buildResult(ctx context.Context, i int, checkResult ocr2keepers.CheckResult, checkResults []ocr2keepers.CheckResult, lookups map[int]*mercury.StreamsLookup) {
+	lookupLggr := s.lggr.With("where", "StreamsLookup")
+	if checkResult.IneligibilityReason != uint8(mercury.MercuryUpkeepFailureReasonTargetCheckReverted) {
+		// Streams Lookup only works when upkeep target check reverts
+		return
+	}
+
+	block := big.NewInt(int64(checkResult.Trigger.BlockNumber))
+	upkeepId := checkResult.UpkeepID
+
+	if s.mercuryConfig.Credentials() == nil {
+		lookupLggr.Errorf("at block %d upkeep %s tries to access mercury server but mercury credential is not configured", block, upkeepId)
+		return
+	}
+
+	// Try to decode the revert error into streams lookup format. User upkeeps can revert with any reason, see if they
+	// tried to call mercury
+	lookupLggr.Infof("at block %d upkeep %s trying to DecodeStreamsLookupRequest performData=%s", block, upkeepId, hexutil.Encode(checkResults[i].PerformData))
+	streamsLookupErr, err := s.packer.DecodeStreamsLookupRequest(checkResult.PerformData)
+	if err != nil {
+		lookupLggr.Debugf("at block %d upkeep %s DecodeStreamsLookupRequest failed: %v", block, upkeepId, err)
+		// user contract did not revert with StreamsLookup error
+		return
+	}
+	streamsLookupResponse := &mercury.StreamsLookup{StreamsLookupError: streamsLookupErr}
+
+	if len(streamsLookupResponse.Feeds) == 0 {
+		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonInvalidRevertDataInput)
+		lookupLggr.Debugf("at block %s upkeep %s has empty feeds array", block, upkeepId)
+		return
+	}
+
+	// mercury permission checking for v0.3 is done by mercury server
+	if streamsLookupResponse.IsMercuryV02() {
+		// check permission on the registry for mercury v0.2
+		opts := s.buildCallOpts(ctx, block)
+		if state, reason, retryable, allowed, err := s.allowedToUseMercury(opts, upkeepId.BigInt()); err != nil {
+			lookupLggr.Warnf("at block %s upkeep %s failed to query mercury allow list: %s", block, upkeepId, err)
+			checkResults[i].PipelineExecutionState = uint8(state)
+			checkResults[i].IneligibilityReason = uint8(reason)
+			checkResults[i].Retryable = retryable
+			return
+		} else if !allowed {
+			lookupLggr.Debugf("at block %d upkeep %s NOT allowed to query Mercury server", block, upkeepId)
+			checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonMercuryAccessNotAllowed)
+			return
+		}
+	} else if streamsLookupResponse.IsMercuryVersionUnkown() {
```

Comment:
as per current implementation this is not really an unknown check but a non  v0.3 check

Reply:
yes, true, will update.